programs/loft.sc

```diff
@@ -0,0 +1,356 @@
+__config() -> {
+    'stay_loaded'->true,
+    'commands'->{
+        'wand'->['wand'],
+        'paste <Block> <replace>'->['paste'],
+        'paste <Block>'->['paste',null],
+        'paste'->['paste','diamond_block',null],
+        'undo'->['undo']
+    },
+    'arguments'->{
+        'Block'->{'type'->'term', 'suggest'->[
+			'white_concrete','yellow_concrete','orange_concrete','purple_concrete',
+			'lime_concrete','red_concrete','green_concrete','magenta_concrete',
+			'brown_concrete','gray_concrete','light_gray_concrete','blue_concrete',
+			'light_blue_concrete','pink_concrete','cyan_concrete','black_concrete'
+			]
+		},
+        'replace'->{'type'->'term','suggest'->['air','water','lava']}
+    }
+};
+
+global_tool='phantom_membrane';
+global_block='diamond_block';
+global_replace=null;
+global_sets=[]; //array containing current information about the m splines created
+global_points=[]; //vector of maximum n points of the current spline that is being created by the player
+global_cset=[]; //grid of nxm points created based on global_sets by the function _normalise_set()
+global_a=[]; //grid of a coefficients needed for bezier surfaces
+global_b=[]; //grid of b coefficients needed for bezier surfaces
+global_c=[]; //grid of c coefficients needed for bezier surfaces
+global_affected_block=[]; //vector containing information about block that the script changed in the last paste operation
+global_history=[]; //vector containing information about block that the script changed in the last N paste operation
+
+wand()->(
+    print('the wand is:');
+    print(global_tool);
+);
+
+__on_player_breaks_block(player, block)->( //starts a new spline
+    item_tuple = query(player,'holds',hand='mainhand'):0;
+    if(item_tuple==global_tool,
+        schedule(0, _(outer(block)) -> set(pos(block), block));
+        _draw_surf(1);
+        global_points=[];
+        global_points+=pos(block);
+        global_sets:(length(global_sets))=global_points;
+        _normalise_set();
+        _draw_surf(10000)
+    );
+    if(item_tuple!=global_tool,
+        if(global_sets!=[],
+            _draw_surf(1);
+            print('operation cancelled');
+            global_sets=[];
+            global_cset=[];
+            global_points=[]
+        )
+    )
+);
+
+__on_player_right_clicks_block(player, item_tuple, hand, block, face, hitvec)->( //expands the spline adding new points
+    if((get(item_tuple,0)==global_tool)&&(hand=='mainhand')&&(global_points!=[]),
+        _draw_surf(1);
```

Comment:
if this call here here is just so that you cancel the old shape and then can call a new one, this would be immediately solved by the suggestion above regarding schedule.